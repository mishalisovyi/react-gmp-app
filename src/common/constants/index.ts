export const Icon = {
  VERTICAL_ELLIPSIS: String.fromCharCode(8942),
  MULTIPLICATION_X: String.fromCharCode(10005),
  TELEPHONE_RECORDER: String.fromCharCode(8981),
};

export const API_URL = 'http://localhost:4000';

export const HTTP_RESPONSE_STATUS_NO_CONTENT = 204;
