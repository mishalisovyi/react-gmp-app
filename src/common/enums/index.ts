export { ButtonType } from './button-type.enum';
export { HttpMethod } from './http-method.enum';
