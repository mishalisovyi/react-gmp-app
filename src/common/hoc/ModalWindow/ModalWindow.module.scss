@use "../../../colors";
@use "../../../mixins";

.ModalWindow {
  &__container {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    z-index: 1;
  }

  &__backdrop {
    width: 100%;
    height: 100%;
    background-color: colors.$young-night;
    opacity: 0.7;
  }

  &__window {
    min-width: 550px;
    min-height: 250px;
    padding: 18px 56px;
    display: flex;
    flex-direction: column;
    box-sizing: border-box;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    background-color: colors.$young-night;
    box-shadow: 0px 0px 10px colors.$black-panther;

    .window {
      &__close-button-wrapper {
        position: absolute;
        top: 24px;
        right: 24px;
      }

      &__title {
        @include mixins.typography-heading;
      }

      &__body {
        flex-grow: 1;
      }

      &__action-buttons {
        display: flex;
        justify-content: flex-end;
        column-gap: 12px;
        margin-top: 16px;
      }
    }
  }
}
