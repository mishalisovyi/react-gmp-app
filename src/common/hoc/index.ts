export { withPortalRendering } from './withPortalRendering/withPortalRendering';
export { ModalWindow } from './ModalWindow/ModalWindow';
