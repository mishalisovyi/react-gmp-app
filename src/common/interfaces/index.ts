export { EnumerableComponentProps } from './EnumerableComponentProps';
export { ModalWindowProps, ModalWindowActionButtonProps } from './ModalWindowProps';
export { SelectListData, SelectListValue } from './SelectListData';
