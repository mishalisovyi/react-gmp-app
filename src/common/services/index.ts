export {
  sendGetRequest,
  sendPostRequest,
  sendPutRequest,
  sendDeleteRequest,
} from './http/http.service';
