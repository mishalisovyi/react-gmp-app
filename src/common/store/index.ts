export * from './actions';
export * from './models';
export { commonReducer } from './reducer';
