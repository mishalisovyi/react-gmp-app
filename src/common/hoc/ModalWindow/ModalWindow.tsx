import React, { useMemo } from 'react';

import { Icon } from 'common/constants';
import { ButtonType } from 'common/enums';
import { ModalWindowProps } from 'common/interfaces';
import { withPortalRendering } from 'common/hoc';
import { Button } from 'common/ui';

import styles from './ModalWindow.module.scss';

const defaultActionButtonAttributes = {
  show: false,
  text: '',
  onClick: () => { },
};

function ModalWindowComponent({
  title = '',
  primaryButton = defaultActionButtonAttributes,
  secondaryButton = defaultActionButtonAttributes,
  onClose,
  children,
}: ModalWindowProps) {
  const primaryButtonElement = useMemo(() => {
    return primaryButton.show
      ? (
        <Button
          type={ButtonType.Primary}
          text={primaryButton.text}
          onClick={primaryButton.onClick}
        />
      )
      : null;
  }, [primaryButton.show, primaryButton.text, primaryButton.onClick]);

  const secondaryButtonElement = useMemo(() => {
    return secondaryButton.show
      ? (
        <Button
          type={ButtonType.Secondary}
          text={secondaryButton.text}
          onClick={secondaryButton.onClick}
        />
      )
      : null;
  }, [secondaryButton.show, secondaryButton.text, secondaryButton.onClick]);

  return (
    <div className={styles['ModalWindow__container']}>
      <div className={styles['ModalWindow__backdrop']} />
      <div className={styles['ModalWindow__window']}>
        <div className={styles['window__close-button-wrapper']}>
          <Button type={ButtonType.Transparent} text={Icon.MULTIPLICATION_X} onClick={onClose} />
        </div>
        <header className={styles['window__title']}>{title}</header>
        <div className={styles['window__body']}>{children}</div>
        <div className={styles['window__action-buttons']}>
          {secondaryButtonElement}
          {primaryButtonElement}
        </div>
      </div>
    </div>
  );
}

export const ModalWindow = withPortalRendering(ModalWindowComponent);
