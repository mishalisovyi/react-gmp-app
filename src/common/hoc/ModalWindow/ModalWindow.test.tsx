import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Icon } from 'common/constants';

import { ModalWindow } from './ModalWindow';

const modalWindowCloseHandler = jest.fn();
const primaryButtonClickHandler = jest.fn();
const secondaryButtonClickHandler = jest.fn();

const primaryButtonData = {
  show: true,
  text: 'Submit',
  onClick: primaryButtonClickHandler,
};

const secondaryButtonData = {
  show: true,
  text: 'Reset',
  onClick: secondaryButtonClickHandler,
};

beforeAll(() => {
  ReactDOM.createPortal = jest.fn((element) => {
    return element;
  }) as any;
});

afterAll(() => {
  (ReactDOM.createPortal as jest.Mock).mockClear();
});

it('should render modal window without buttons', () => {
  const tree = renderer.create(
    <ModalWindow title="test 1" onClose={modalWindowCloseHandler}>
      <p>test</p>
    </ModalWindow>,
  );

  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render modal window with only primary button', () => {
  const tree = renderer.create(
    <ModalWindow title="test 2" primaryButton={primaryButtonData} onClose={modalWindowCloseHandler}>
      <p>test</p>
    </ModalWindow>,
  );

  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render modal window with primary and secondary buttons', () => {
  const tree = renderer.create(
    <ModalWindow title="test 3" primaryButton={primaryButtonData} secondaryButton={secondaryButtonData} onClose={modalWindowCloseHandler}>
      <p>test</p>
    </ModalWindow>,
  );

  expect(tree.toJSON()).toMatchSnapshot();
});

it('should handle primary button click', async () => {
  render(
    <ModalWindow title="test 4" primaryButton={primaryButtonData} secondaryButton={secondaryButtonData} onClose={modalWindowCloseHandler}>
      <p>test</p>
    </ModalWindow>,
  );

  await userEvent.click(screen.getByText(primaryButtonData.text));

  expect(primaryButtonClickHandler).toHaveBeenCalled();
});

it('should handle secondary button click', async () => {
  render(
    <ModalWindow title="test 5" primaryButton={primaryButtonData} secondaryButton={secondaryButtonData} onClose={modalWindowCloseHandler}>
      <p>test</p>
    </ModalWindow>,
  );

  await userEvent.click(screen.getByText(secondaryButtonData.text));

  expect(secondaryButtonClickHandler).toHaveBeenCalled();
});

it('should handle modal window closing', async () => {
  render(
    <ModalWindow title="test 6" primaryButton={primaryButtonData} secondaryButton={secondaryButtonData} onClose={modalWindowCloseHandler}>
      <p>test</p>
    </ModalWindow>,
  );

  await userEvent.click(screen.getByText(Icon.MULTIPLICATION_X));

  expect(modalWindowCloseHandler).toHaveBeenCalled();
});
