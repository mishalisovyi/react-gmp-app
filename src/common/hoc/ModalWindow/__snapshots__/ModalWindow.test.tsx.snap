// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render modal window with only primary button 1`] = `
<div>
  <div />
  <div>
    <div>
      <button
        className="button button--transparent"
        onClick={[MockFunction]}
        type="button"
      >
        ✕
      </button>
    </div>
    <header>
      test 2
    </header>
    <div>
      <p>
        test
      </p>
    </div>
    <div>
      <button
        className="button button--primary"
        onClick={[MockFunction]}
        type="button"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;

exports[`should render modal window with primary and secondary buttons 1`] = `
<div>
  <div />
  <div>
    <div>
      <button
        className="button button--transparent"
        onClick={[MockFunction]}
        type="button"
      >
        ✕
      </button>
    </div>
    <header>
      test 3
    </header>
    <div>
      <p>
        test
      </p>
    </div>
    <div>
      <button
        className="button button--secondary"
        onClick={[MockFunction]}
        type="button"
      >
        Reset
      </button>
      <button
        className="button button--primary"
        onClick={[MockFunction]}
        type="button"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;

exports[`should render modal window without buttons 1`] = `
<div>
  <div />
  <div>
    <div>
      <button
        className="button button--transparent"
        onClick={[MockFunction]}
        type="button"
      >
        ✕
      </button>
    </div>
    <header>
      test 1
    </header>
    <div>
      <p>
        test
      </p>
    </div>
    <div />
  </div>
</div>
`;
